### PR TITLE
🎉  Homepage redesign

### PR DESCRIPTION
**Pull request by ikesau**
[Figma](https://www.figma.com/file/m4MLbbAmIiQLwPSG1OtA0I/%5BProject%5D-Homepage?type=design&node-id=746-34381&mode=design&t=I3M2pjfD8AAsnPbq-0)

[Test document](https://docs.google.com/document/d/1-X5zMYoLTJqD6kidXYd1GuIDj1uhlPe5-Em5B7vsgNo/edit)

There's a big change in bundle size because the new `homepage-search` component moves the algolia library into `common.mjs`

The `FrontPage` code is all still in here (with a few modifications where I overrode/reused/shared CSS) but will be removed once this branch is merged and we've published the new homepage.

## New Features
- A new `homepage` gdoc type
  - It's basically a body of new, homepage-specific components and then some static content appended beneath
  - If a published `homepage` exists, it will be baked as `index.html`, otherwise we bake the default frontpage.
- A new `homepageMetadata` attachments context (to provide stats on how many charts and topics we have)

## New components
- `pill-row`
- `homepage-search`
- `homepage-intro`
- `latest-data-insights`

## TODO
- Design fixes
- Accordion for all our topics list

## Summary by CodeRabbit

- **New Features**
	- Introduced support for the `Homepage` document type across the platform, including icons, settings, and preview functionalities.
	- Added new homepage-specific content blocks such as `PillRow`, `HomepageIntro`, `HomepageSearch`, and `LatestDataInsightsBlock`.
	- Implemented a new layout and components for displaying the latest data insights on the homepage.
	- Enhanced search functionality with additional props for customization and improved user interaction.
- **Style**
	- Updated styles across various components and pages, including color adjustments and grid layout changes to support new homepage elements.
	- Introduced new SCSS files for styling latest data insights and homepage-specific components.
- **Documentation**
	- Added structured data for Google search actions on the `SearchPage`.
- **Refactor**
	- Centralized database operations related to data insights into the `db` module.
	- Refined the handling and rendering logic for various content types, including the homepage and data insights.
- **Chores**
	- Updated imports and variable names to reflect changes in content structure and types.

### CODE REVIEW

File: site/gdocs/components/HomepageIntro.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import { groupBy } from "@ourworldindata/utils"
+import { useLinkedDocument } from "../utils.js"
+import { DocumentContext } from "../OwidGdoc.js"
+import Image from "./Image.js"
+import { BlockErrorFallback } from "./BlockErrorBoundary.js"
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
+import { faArrowRight } from "@fortawesome/free-solid-svg-icons"
+
+type FeaturedWorkTileProps =
+    EnrichedBlockHomepageIntro["featuredWork"][number] & {
```
Comment (mlbrgl) [inline site/gdocs/components/HomepageIntro.tsx:13]: nit: `EnrichedBlockHomepageIntroPost`?

**Review by mlbrgl — COMMENTED** [review]
(commit of in-progress review comments)

File: db/db.ts [hunk 2]
```diff
@@ -204,3 +208,63 @@ export const getPublishedExplorersBySlug = async (
...
+
+export const getLatestDataInsights = (
+    limit = 5
+): Promise<MinimalDataInsightInterface[]> => {
+    return knexRaw(
+        `
+        SELECT
+            content->>'$.title' AS title,
+            publishedAt,
+            ROW_NUMBER() OVER (ORDER BY publishedAt DESC) - 1 AS \`index\`
```
Comment (mlbrgl) [inline db/db.ts:220]: seems a bit heavy handed, given that we can get the index in the `map()` below (EDIT: I'm seeing now this has been merely moved so not really part of this PR)

File: db/model/Gdoc/rawToEnriched.ts [hunk 3]
```diff
@@ -2056,3 +2069,177 @@ function parseKeyIndicatorCollection(
...
+            title: "",
+        }
+    }
+
+    if (!raw.value.title) {
+        return createError({
+            message: "Pill row is missing a title",
+        })
+    }
+    if (!raw.value.pills) {
```
Comment (mlbrgl) [inline db/model/Gdoc/rawToEnriched.ts:2088]: we might want on the `length` here too

**Review by mlbrgl — APPROVED** [review]
it's really nice to see how the gdoc architecture scales to this special use case - to the point where it is not that special any more, and we barely needed a template for it.

I also like how you thoroughly handle edge cases and error reporting 👍

File: db/db.ts [hunk 4]
```diff
@@ -204,3 +208,63 @@ export const getPublishedExplorersBySlug = async (
...
+        FROM posts_gdocs
+        WHERE content->>'$.type' = '${OwidGdocType.DataInsight}'
+            AND published = TRUE
+            AND publishedAt < NOW()
+        ORDER BY publishedAt DESC
+        LIMIT ?
+        `,
+        knexInstance(),
```
[the remark this answers is not in the document]
Comment (ikesau) [reply db/db.ts:228]: Just making sure I understand you - you mean all these db functions should take knex as an argument and we pass the same instance around everywhere? 🙂

File: site/css/grid.scss [hunk 5]
```diff
@@ -61,7 +61,7 @@ $grid-responsive: (lg, md, sm);
 // Use this when you want a full-page-width grid container
 // 1 (auto) | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 (auto)
 .grid-cols-12-full-width {
-    --outer-column-width: calc((50vw - 640px - var(--grid-gap)));
+    --outer-column-width: calc((50vw - 647.5px - var(--grid-gap)));
```
[the remark this answers is not in the document]
Comment (ikesau) [reply site/css/grid.scss:64]: I actually can't 😭 

I have no idea why it's happening but it seems like every couple of months this rule stops working and the 12-column grid is no longer 1280px wide.

load the homepage and look at `.latest-data-insights-block__card-container` on a desktop - it's 1280px

then change `--outer-column-width` to what is was before and the element's 1265px wide

but last time I did this I swear it was too large so I bumped it back down. 😭😭

File: site/gdocs/pages/Homepage.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
```
[the remark this answers is not in the document]
Comment (ikesau) [reply site/gdocs/pages/Homepage.tsx:1]: Wow! Good to know!

File: db/db.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlbrgl) [reply db/db.ts:228]: yep 😉

File: site/css/grid.scss (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlbrgl) [reply site/css/grid.scss:64]: interesting I'm seeing different widths on Chrome:
- 647.5 -> 1295px
- 640 -> 1280px

**Pull request merged**